lib/improver/utilities/cube_units.py

```diff
@@ -32,14 +32,125 @@
 
 import numpy as np
 
+import iris
 from iris.exceptions import CoordinateNotFoundError
 
 from improver.units import DEFAULT_UNITS
 
 
+def enforce_units_and_dtypes(cubes, coords=None, enforce=True):
+    """
+    Function to check the units and datatypes of cube diagnostics and
+    coordinates against the manifest in improver.units, with option to
+    enforce or fail for non-conforming data.
+
+    Args:
+        cubes (iris.cube.Cube or iris.cube.CubeList):
+            Cube or list of cubes to be checked
+
+    Kwargs:
+        coords (list or None):
+            List of coordinate names to check.  If None, checks all
+            coordinates present on the input cubes.
+        enforce (bool):
+            If True, this function returns a list of conformant cubes.
+            If False, a ValueError is thrown if the cubes do not conform.
+
+    Raises:
+        ValueError: if "enforce=False" and the input cubes do not conform
+            to the datatypes and units standard.
+
+    Returns:
+        new_cubes (iris.cube.CubeList):
+            New cubelist with conformant datatypes and units
+    """
+    # convert input to CubeList
+    if isinstance(cubes, iris.cube.Cube):
+        cubes = [cubes]
+
+    # make a copy of the input cube list
+    new_cubes = enforce_diagnostic_units_and_dtypes(cubes, inplace=False)
+    # set up coordinates to check
+    if coords is not None:
+        all_coords = list(set(coords))
+    else:
+        all_coords = []
+        for cube in cubes:
+            all_coords.extend(
+                [coord.name() for coord in new_cubes[0].coords()])
```

Comment:
I am expecting the list of input cubes to have potentially different sets of coordinates, hence the loop.  I'm thinking in terms of eg the orographic enhancement plugin, where you might want just to pump everything into this function and get sensible answers out, despite the fact that they are different diagnostics.  Fundamentally, I think we should either operate on a single cube, or if on a cube list, be prepared to accept whatever heterogeneity is permitted in an iris `CubeList` object (which is basically anything).  What do you think?

On the bug: thanks for the spot.  I've fixed this and will make sure the unit tests cover it.

Reply:
OK, I can see that this might be applied wholesale across a range of cubes, that might potentially have different coordinates, so yes, using the `for` loop is fine.